### PR TITLE
[react-native-auth0] add type definition for custom claims in user information

### PR DESCRIPTION
**Pull request by marknelissen**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [custom claims](https://auth0.com/docs/scopes/current/custom-claims)
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/react-native-auth0/index.d.ts [hunk 1]
```diff
@@ -75,15 +76,18 @@ export interface ResetPasswordParams {
...
     email: string;
     emailVerified: boolean;
     name: string;
     nickname: string;
     picture: string;
     sub: string;
     updatedAt: string;
-}
+} & {
+    [key: string]: string;
```
Comment (ascariandrea) [inline types/react-native-auth0/index.d.ts]: Could maybe make more sense to generalize `UserInfo` to accept a type of the custom claims? That would make possible to define a meaningful type and export it around the app

Comment (marknelissen) [reply types/react-native-auth0/index.d.ts]: I don't really know how that would work. The claims are part of the actual properties of the UserInfo, not contained within a single property, eg.
`{
  "email": "foo@bar.org",
  "emailVerified": true,
  "https://acme.com/foo": "bar"
}`

The only thing that differentiates them from the other properties, is that the key should always be a valid url format, using http or https protocol.

Normally I would just have added the `[key: string]: string;` next to the other properties of the UserInfo interface, but this is not allowed due to not all properties being of type string.

Maybe this isn't an issue, and I am wrong with limiting the values of the claims to strings, but that is the only thing I tried until now.

Comment (ascariandrea) [reply types/react-native-auth0/index.d.ts]: In this case, I think generics address your problem:
you can define the `UserInfo` to accept a *type parameter*, i.e. `type UserInfo<Claims = {}> = {email: string, emailVerified: boolean} & Claims`.
In this way you can extend the base `UserInfo` with your *claims*.

Does it make sense?

Can you give it a try?

Comment (ascariandrea) [reply types/react-native-auth0/index.d.ts]: I would go for the generic implementation as the *type parameter* is defaulted to `{}` so is an opt-in for the users who want more type safety, while it stays simple for those who need just the plain `UserInfo`.

The comment could be helpful, but I think if deductible from the type name - being in the `react-native-auth0` context

File: types/react-native-auth0/index.d.ts [hunk 2]
```diff
@@ -75,27 +76,28 @@ export interface ResetPasswordParams {
...
     exchange(params: ExchangeParams): Promise<string>;
     logoutUrl(params: LogoutParams): string;
     passwordRealm(params: PasswordRealmParams): Promise<PasswordRealmResponse>;
-
     refreshToken(params: RefreshTokenParams): Promise<any>;
     resetPassword(params: ResetPasswordParams): Promise<any>;
     revoke(params: RevokeParams): Promise<any>;
-    userInfo(params: UserInfoParams): Promise<UserInfo>;
+    /* tslint:disable-next-line no-unnecessary-generics */
+    userInfo<CustomClaims>(params: UserInfoParams): Promise<UserInfo<CustomClaims>>;
```
Comment (ascariandrea) [inline types/react-native-auth0/index.d.ts]: Add a default for `CustomClaims` here and I'm happy with this PR

**Review by ascariandrea — CHANGES_REQUESTED** [review]

**Review by ascariandrea — APPROVED** [review]

**Pull request merged**